### PR TITLE
feat: support highlighting tokens by range

### LINKED ISSUE #224: Respecting LSP hover ranges
lsif-server returns ranges alongside hover text, but these are not used to adjust the highlighted token, which can appear incorrect due to the current heuristic scan.

```json
{
  "hover": {
    "markdown": {
      "text": "```go\npackage \"golang.org/x/tools/go/ast/astutil\"\n```
\n\n---\n\nPackage astutil contains common utilities for working with the Go AST.”
    },
    “range”: {
      “start”: {
        “line”: 12,
        “character”: 2
      },
      “end”: {
        “line”: 12,
        “character”: 35
      }
    }
  }
}
```

This is particular noticeable with go import paths, which should highlight the entire string content, not just the first ident-like token:

Screen Shot 2020-02-27 at 10 06 29 AM

### PR DESCRIPTION
**Pull request by tjkandala**
Closes #224

Find and highlight tokens by range, not just position.

Limitations:
- `start` and `end` position must be on the same line
- Highlight class will be applied to all elements within the range

### CODE REVIEW

File: src/hoverifier.ts [hunk 1]
```diff
@@ -789,10 +789,6 @@ export function createHoverifier<C extends object, D, A>({
                         })
                     }
 
-                    // The requested position is is 0-indexed; the code here is currently 1-indexed
-                    const { line, character } = pos
-                    pos = { line: line + 1, character: character + 1, ...pos }
```
Comment (felixfbecker) [inline src/hoverifier.ts:794]: **question**: Why was this removed? This is passed back into user-provided functions like `adjustPosition`, so won't this cause problems / is this not a breaking change?

File: src/token_position.ts [hunk 2]
```diff
@@ -267,64 +268,109 @@ export function findElementWithOffset(
...
-    let targetNode: Node | undefined
-    for (const [i, node] of textNodes.entries()) {
-        const text = node.textContent ?? ''
-        if (offsetStep <= offset && offsetStep + text.length > offset) {
-            targetNode = node
-            nodeIndex = i
-            break
+    let startNode: Node
+    let endNode: Node
+    if (typeof offsetEnd === 'number' && offsetEnd > offsetStart) {
```
Comment (felixfbecker) [inline src/token_position.ts:273]: **question**: why do ranges and positions need to be handled in completely separate branches?

**Review by felixfbecker — COMMENTED** [review]
Nice! Codecov shows there are a couple of branches in the new code never hit by tests (decreasing coverage). Are they testable?

File: src/hoverifier.ts (hunk 1 above)
Comment (tjkandala) [reply src/hoverifier.ts:794]: The incremented `line` and `character` were overwritten by spreading `pos`. 

Also, it seems that `PositionAdjuster` is meant to receive and return a 0-based position. Furthermore, the position's line and character are incremented by 1 in the next `switchMap` in `getHighlightedRange` anyways. If my assumptions are correct, then there's no need to convert the indexing at this point. If not, then the fix is to spread `pos` before incrementing line and character.

File: src/token_position.ts (hunk 2 above)
Comment (tjkandala) [reply src/token_position.ts:273]: There's not much overlap between these branches. When `findElementWithOffset` is called with a `Position`, we should look forwards and backwards to find all elements that comprise the token at that position. When it's called with a `Range` we assume that the position boundaries are intentional (i.e. that the user either already knows the start and end positions of the token, or doesn't want to highlight a discrete token)

**Comment by tjkandala** [comment]
I don't see much new _logic_ that's untested:
- The case where `getCodeElementFromLineNumber` returns `undefined` wasn't tested before, but is now duplicated due to branching on `positionOrRange`, which adds an "untested" line. 
- The case where `isFirstCharacterDiffIndicator` is provided and returns true for the code element wasn't tested before, but is now duplicated due to branching on `positionOrRange`, which adds an "untested" line.
- The whole `hoverOverlayRerenders` subscription is untested.

I'll add tests for invalid ranges. Should we also test the previously untested cases?

File: src/hoverifier.ts (hunk 1 above)
Comment (felixfbecker) [reply src/hoverifier.ts:794]: Oh haha, you're right.

**Review by felixfbecker — APPROVED** [review]
@marekweb could you review too?

**Review by felixfbecker — APPROVED** [review]
